### DISCUSSION TITLE
Normalization differences between spatial and scRNAseq dataset and model adjustment using only one spatial dataset

### DISCUSSION THREAD

**Discussion by Chuang1118** [BayraktarLab/cell2location#30] [Q&A]
Dear Author,

First, thank you for this great tool and this new API
I'm very interested in applying the cell2location method to my analysis.
I already have an integrated seurat object (15 integrated samples) that I have generated through seurat standard workflow (i.e normalization method = log-normalization) that will be used as the reference (To note, this reference dataset have cell types annotation).
In this object, I have 2 assays : RNA and integrated (I don't have SCT assay).

I also have one spatial dataset ( a seurat object) as query data.
In this spatial object, I have used sctransform normalization. Thus, it has 2 assays : Spatial and SCT.

Now, I want to link this 2 datasets : spatial dataset and scRNA-seq dataset. 
Here, spatial object has SCT slot while scRNA-seq object has not (as both being normalized using 2 different methods).

Just a few beginner's question.
In the part of Model-based estimation of reference expression signatures of cell types (1/3).
1/ Reference data must have SCT assay? According to what I have understood, the model use raw counts (thus not having SCT slot in the reference should not be a problem).

In Spatially mapping cell types(2/3), you have selected 2 Visium sections to speed up the analysis.
2/ I have only one Visium, Is this a problem?
3/ The overall question is : Is cell2loction adapted to my situation? 

I'm a beginner in Scanpy.
4/ If yes for question 3. how to pass Seurat object to Anndata.

Thanks in advance
Kind regards,
Chuang

**Comment by vitkl** [comment]
Hi Chuang

Thanks for using our package! I think your message belongs more in GitHub discussions rather than issues.

1. You need 3 things: counts for reference data, cell annotations for reference data and counts for spatial data.

2. You can use cell2location on one and on many Visium samples, although it is beneficial to use many samples.
3. Yes :)

4. This is a good question. I generally need to convert anndata to Seurat objects rather that Seurat to anndata objects, but I found this package from Seurat authors very useful: https://github.com/mojaveazure/seurat-disk, https://rdrr.io/github/mojaveazure/seurat-disk/man/Convert.html

I hope this helps - don't hesitate to get back.

**Comment by Chuang1118** [comment]
Hello Author,

Now, I'm in step 2(cell2location.run_cell2location) using my dataset.
I have error message like in issue #31. 
I have run the same code in docker and singluarity.
In the docker (mode CPU in local) works, but not in singluarity (mode GPU in HPC Slurm).

In my opinion, singularity and docker aren't same version.
Because, in docker version,  summ_sc_data_args['selection']  hasn't  'selection': "cluster_specificity".
Only available: None, high_cv, cluster_markers and AutoGeneS

Question:
1/ Could you talk me which tutorial code you work with singularity ?

2/ I have also this error / Warning, I have used cuda/10.1, In the HPC doesn't have 10.2.
```
Can not use cuDNN on context None: cannot compile with cuDNN. We got this error:
b'In file included from /tmp/try_flags_hj9m6ulp.c:4:0:\n/usr/include/cudnn.h:63:10: fatal error: driver_types.h: No such file or directory\n #include "driver_types.h"\n          ^~~~~~~~~~~~~~~~\ncompilation terminated.\n'
Mapped name None to device cuda: Tesla P100-PCIE-12GB (0000:02:00.0)  
```

3/ Can you update singularity image? 


Code:
```
r= cell2location.run_cell2location(
      sc_data=inf_aver / adata_snrna_raw,
      sp_data=adata_vis,
      train_args={'sample_name_col': 'sample'}
) 
```
 Summarising single cell clusters 
 Creating model  - time 0.01 min
Analysis name: LocationModelLinearDependentW_21clusters_1665locations_21746genes
 Training model 
 0.01% [1/20000 00:04<26:08:19 Average Loss = 9.6718e+07]

In Slurm :
ImportError: libpython3.7m.so.1.0: cannot open shared object file: No such file or directory

Thank you in advance

**Reply by vitkl** [reply]
2/ I would recommend using `summ_sc_data_args['selection'] = None` with a total number of genes between 8-16k rather than doing gene selection, especially when mapping a lot of cell types (e.g. >50) and when the spatial data quality is not great. 

3/ I will update the singularity image next week. We are working on other features too.

**Comment by vitkl** [comment]
Looks like 1 other person is reporting the issue about Slurm too. I will reply to other points next week.

**Comment by Chuang1118** [comment]
Hello Vitkl,

After the parameters optimization on HPC (Slurm).
I confirm that running step2 (cell2location.run_cell2location) using your singularity (cell2location-v0.05-alpha.sif) in mode GPU without interactive mode. 
It really does work well on Slurm system with GPU.

Now, I've finished all steps using my dataset. I will talk about the results with my boss.
If we get some very interesting results, without doubt, we will cite your paper. I hope so !
You can close this thread then, no further questions for now :)
Thanks again and great work !

Chuang

Here is my submitting Code
```
#!/bin/sh

#SBATCH -J c2l
#SBATCH -A b098
#SBATCH --nodes=1
#SBATCH --mem=50G
#SBATCH -t 02:30:00
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=dong@ciml.univ-mrs.fr
#SBATCH -p gpu 
#SBATCH --gres=gpu:2

module load userspace/all
module load singularity/3.5.1

hostname

singularity exec \
        --nv \
        --no-home  \
        -B /scratch/cdong:/scratch/cdong ./simg/cell2location-v0.05-alpha.sif \
/bin/bash -c "HOME=$(mktemp -d) jupyter nbconvert --to notebook --execute ./notebooks/analysis_1_spatial_mapping.ipynb --output=/scratch/cdong/cell2loc/fl/notebooks/out_refFL.ipynb"
```

**Reply by vitkl** [reply]
Hi @Chuang1118

Great to hear that you sorted this issue out! Could you please highlight which specific change fixed this for you? This could be quite useful for other Slurm users. Thanks!

Good luck with the analysis and happy to help!
